ethers-contract/ethers-contract-abigen/src/contract.rs

```diff
@@ -227,7 +230,7 @@ impl Context {
             .map(|derive| syn::parse_str::<Path>(derive))
             .collect::<Result<Vec<_>, _>>()
             .context("failed to parse event derives")?;
-
+        let abi_str = serde_json::to_string(&abi).context("fail to serialize abi to json")?;
```

Comment:
this is a bit wasteful because we only have to do this in the else branch were we try to deserialize the `Contract {abi}`, so instead we should make `abi_str` a `mut` and simply update it if we indeed have a hh style artifact.

Also this is not quite optimal

https://github.com/gakonst/ethers-rs/blob/84408961fc9a4e1d37954295db32dd1e885f6047/ethers-contract/ethers-contract-abigen/src/contract.rs#L163-L167

if we deserialize in both branches we can get rid of the `abi_str.clone()`